linera-service/src/chain_listener.rs

```diff
@@ -46,55 +74,135 @@ where
     S: Store + Clone + Send + Sync + 'static,
     ViewError: From<S::ContextError>,
 {
-    /// Creates a new chain listener given a client chain.
-    pub fn new(config: ChainListenerConfig, client: Arc<Mutex<ChainClient<P, S>>>) -> Self {
-        Self { config, client }
+    /// Creates a new chain listener given client chains.
+    pub(crate) fn new(config: ChainListenerConfig, clients: ChainClients<P, S>) -> Self {
+        Self { config, clients }
     }
 
     /// Runs the chain listener.
-    pub async fn run<C>(self, mut context: C) -> Result<(), anyhow::Error>
+    pub async fn run<C>(self, mut context: C, storage: S) -> Result<(), anyhow::Error>
     where
         C: ClientContext<P>,
     {
-        let mut notification_stream = ChainClient::listen(self.client.clone()).await?;
-        let mut tracker = NotificationTracker::default();
-        while let Some(notification) = notification_stream.next().await {
-            if tracker.insert(notification.clone()) {
-                info!("Received new notification: {:?}", notification);
-                if self.config.delay_before_ms > 0 {
-                    tokio::time::sleep(Duration::from_millis(self.config.delay_before_ms)).await;
-                }
-                {
-                    let mut client = self.client.lock().await;
-                    match &notification.reason {
-                        Reason::NewBlock { .. } => {
-                            if let Err(e) = client.update_validators().await {
-                                warn!(
-                                    "Failed to update validators about the local chain after \
-                                    receiving notification {:?} with error: {:?}",
-                                    notification, e
-                                );
-                            }
-                        }
-                        Reason::NewIncomingMessage { .. } => {
-                            if let Err(e) = client.process_inbox().await {
-                                warn!(
-                                    "Failed to process inbox after receiving new message: {:?} \
-                                    with error: {:?}",
-                                    notification, e
-                                );
-                            }
-                        }
+        let mut streams = HashMap::new();
+        self.update_streams(&mut streams, &mut context, &storage)
+            .await?;
+
+        while let (Some((notification, client)), _, _) =
+            future::select_all(streams.values_mut().map(StreamExt::next)).await
```

Comment:
One drawback of using this approach (with `select_all` or `select!`) is that the stream futures must be cancellation-safe. Have you considered using one separate task per stream?